doc/source/development/roadmap.rst

```diff
@@ -0,0 +1,130 @@
+.. _roadmap:
+
+=======
+Roadmap
+=======
+
+This page provides an overview of the major themes pandas' development. Implementation
+of these goals may be hastened with dedicated funding.
+
+Extensibility
+-------------
+
+Pandas Extension Arrays provide 3rd party libraries the ability to
+extend pandas' supported types. In theory, these 3rd party types can do
+everything one of pandas. In practice, many places in pandas will unintentionally
+convert the ExtensionArray to a NumPy array of Python objects, causing
+performance issues and the loss of type information. These problems are especially
+pronounced for nested data.
+
+We'd like to improve the handling of extension arrays throughout the library,
+making their behavior more consistent with the handling of NumPy arrays.
+
+String Dtype
+------------
+
+Currently, pandas stores text data in an ``object`` -dtype NumPy array.
+Each array stores Python strings. While pragmatic, since we rely on NumPy
+for storage and Python for string operations, this is memory inefficient
+and slow. We'd like to provide a native string type for pandas.
+
+The most obvious alternative is Apache Arrow. Currently, Arrow provides
+storage for string data. We can work with the Arrow developers to implement
+the operations needed for pandas users (for example, ``Series.str.upper``).
+These operations could be implemented in Numba (
+as prototyped in `Fletcher <https://github.com/xhochy/fletcher>`__)
+or in the Apache Arrow C++ library.
+
+Apache Arrow Interoperability
+-----------------------------
+
+`Apache Arrow <https://arrow.apache.org>`__ is a cross-language development
+platform for in-memory data. The Arrow logical types are closely aligned with
+typical pandas use cases (for example, support for nullable integers).
+
+We'd like have a pandas DataFrame be backed by Arrow memory and data types
+by default. This should simplify pandas internals and ensure more consistent
+handling of data types through operations.
+
+Block Manager Rewrite
+---------------------
+
+We'd like to replace pandas current internal data structures (a collection of
+1 or 2-D arrays) with a simpler collection of 1-D arrays.
+
+Pandas internal data model is quite complex. A DataFrame is made up of
+one or more 2-dimension "blocks", with one or more blocks per dtype. This
+collection of 2-D arrays is managed by the BlockManager.
+
+The primary benefit of the BlockManager is improved performance on certain
+operations (construction from a 2D array, binary operations, reductions across the columns),
+especially for wide DataFrames. However, the BlockManager substantially increases the
+complexity and maintenance burden of pandas'.
```

Comment:
I know I'm boring. But I think we agree the problem is not the model, it's the implementation. Running operations across 1-d arrays of different dtypes has a complexity which is at least comparable to the BlockManager one. And sure, as of today nobody is rewriting the implementation, but this document is about the future. Couldn't we just say we need to rewrite the internals?

Reply:
(If we do want to be specific, one thing which is sure is that the BlockManager shouldn't see the indexes - see my comment on indexing)

Reply:
Would you care to condense https://github.com/pandas-dev/pandas/wiki/(Tentative)-rules-for-restructuring-indexing-code into a paragraph appropriate for a roadmap?

Reply:
Sure

```
Decoupling of indexing code
--------------------------

The code for getting and setting values in pandas structures needs to be better
organized. In particular, a clear separation must be made between code that
converts indexes (e.g., for ``.loc``) to positions, and code that uses such
positions to get/set values. This is related to the rework of the internals:
the Blockmanager object currently supports, and uses internally even for
positional indexing, key indexing; instead, it should only work with positional
indexing, and the translation of keys to positions should be entirely done at a
higher level.

More details are discussed at https://github.com/pandas-dev/pandas/wiki/(Tentative)-rules-for-restructuring-indexing-code
```

(anybody please feel free to reword, and to remove the link if not necessary)

Reply:
Thanks! Lightly modified version included in c3103704e58706960a1385a595b3098e9e4c5530